server/src/main/java/io/mantisrx/master/jobcluster/job/JobActor.java

```diff
@@ -1364,29 +1378,40 @@ private void initializeRunningWorkers() {
             LOGGER.trace("In initializeRunningWorkers for Job {}", jobId);
             // publish a refresh before enqueuing tasks to the Scheduler, as there is a potential race between
             // WorkerRegistryV2 getting updated and isWorkerValid being called from SchedulingService loop
-            // If worker is not found in the SchedulingService loop, it is conisdered invalid and prematurely
-            // removed from Fenzo state
+            // If worker is not found in the SchedulingService loop, it is considered invalid and prematurely
+            // removed from Fenzo state.
             markStageAssignmentsChanged(true);
 
-            for (IMantisStageMetadata stageMeta : mantisJobMetaData.getStageMetadata().values()) {
+            List<JobWorker> workersToResubmit = new ArrayList<>();
 
+            for (IMantisStageMetadata stageMeta : mantisJobMetaData.getStageMetadata().values()) {
                 Map<Integer, WorkerHost> workerHosts = new HashMap<>();
+
                 for (JobWorker worker : stageMeta.getAllWorkers()) {
                     IMantisWorkerMetadata wm = worker.getMetadata();
+
+                    Optional<WorkerPorts> workerPortsOptional = wm.getPorts();
+                    if (!workerPortsOptional.isPresent() || !workerPortsOptional.get().isValid()) {
+                        numMissingWorkerPorts.increment();
+                        // Mark these workers with missing ports for resubmission.
+                        workersToResubmit.add(worker);
+                    }
+
                     if (WorkerState.isRunningState(wm.getState())) {
                         // send fake heartbeat
                         try {
                             WorkerEvent fakeHB = new WorkerHeartbeat(new Status(jobId.getId(), stageMeta.getStageNum(),
                                     wm.getWorkerIndex(), wm.getWorkerNumber(), Status.TYPE.HEARTBEAT, "",
                                     MantisJobState.Started, System.currentTimeMillis()));
                             worker.processEvent(fakeHB, jobStore);
-                        } catch (Exception e) {
+                        } catch (InvalidWorkerStateChangeException | IOException e) {
                             e.printStackTrace();
```

Comment:
A nitpick but adding job id to the error log may be helpful as the generic stack trace may not reveal where the issue occurred

Reply:
Agreed. This was left from before but I'll go ahead and change it.